Review TLB invalidation sequence

etienne-lms opened the pull request:
These changes reviews the TLB invalidation support and somewhat optimizes the TLB maintenance in the pager to prevent executing to many synchronization barriers when possible.

~~This change also removes the statck unwind from cache maintenance primitives, including data/instr. cache and TLBs.~~ **edited:** good question, why removing the stack unwind? no answer. discard.

etienne-lms commented:
These changes were tested over qemu (armv7 and armv8) and the b2260 (armv7 based board).
I hope to test on fvp (but currently failing to run it since some gcc/libstdc version issues).

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -238,8 +238,7 @@ static void set_alias_area(tee_mm_entry_t *mm)
 
-	/* TODO only invalidate entries touched above */
-	core_tlb_maintenance(TLBINV_UNIFIEDTLB, 0);
+	inval_tlb_for_range(smem, nbytes);
```
This function is only implemented in a later patch.

etienne-lms replied:
yep, thanks: I committed it in the wrong commit...

jenswi-linaro commented on core/arch/arm/include/mm/core_mmu.h:
```diff
@@ -446,7 +446,7 @@ enum teecore_tlb_op {
 };
 
-int core_tlb_maintenance(int op, unsigned int a);
+int core_tlb_maintenance(int op, unsigned long a);
```
Why do we need this function at all?

etienne-lms replied:
Actaully... we don't really need it.
I will remove it...

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -198,6 +199,29 @@ static void pager_unlock(uint32_t exceptions)
 	cpu_spin_unlock_xrestore(&pager_spinlock, exceptions);
 }
 
+static void inval_tlb_for_pages(vaddr_t *va, size_t nb_elts)
```
These two functions should preferably not be located in this file as it's complicated enough as it is.

etienne-lms replied:
Ok, i will move these next to the other tbl inval routines.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -814,15 +835,15 @@ static void rem_area(struct tee_pager_area_head *area_head,
 			pgt_dec_used_entries(area->pgt);
 			pmem->area = NULL;
 			pmem->pgidx = INVALID_PGIDX;
+			break;
```
We need to check all pmem:s as several can belong to the same area.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -954,6 +976,9 @@ static bool tee_pager_unhide_page(vaddr_t page_va)
 			if (pa != get_pmem_pa(pmem))
 				panic("unexpected pa");
 
+			TAILQ_REMOVE(&tee_pager_pmem_head, pmem, link);
```
Why are these two lines moved?

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -1012,10 +1035,16 @@ static void tee_pager_hide_pages(void)
 			a = TEE_MATTR_HIDDEN_BLOCK;
 		area_set_entry(pmem->area, pmem->pgidx, pa, a);
+
+		va_tbl[va_cnt++] = area_idx2va(pmem->area, pmem->pgidx);
```
Normally (except at boot and perhaps some corner case) we'll only hide one page at a time so this is a bit complicated.

etienne-lms replied:
Ok, let's simplify this and invalidate tbls at each pass in the loop.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -1069,11 +1098,11 @@ static struct tee_pager_pmem *tee_pager_get_page(struct tee_pager_area *area)
 		uint32_t a;
 
 		assert(pmem->area && pmem->area->pgt);
+		pgt_dec_used_entries(pmem->area->pgt);
```
Why is this line moved?

etienne-lms replied:
Just I wanted the tlb maintenance to be right after the mmu table update.

jenswi-linaro replied:
That shouldn't matter, this is still supposed to be atomic for the other cores. I'm OK with changing though.

etienne-lms commented on core/arch/arm/include/arm32.h:
```diff
@@ -368,6 +368,16 @@ static inline void dsb(void)
 	asm volatile ("dsb");
 }
 
+static inline void dsbish(void)
```
ok

etienne-lms commented on core/arch/arm/kernel/ssvce_a32.S:
```diff
@@ -49,61 +49,50 @@
+	write_tlbiallis 	/* Invalidate TLBs */
+	dsb	ish		/* Sync with tlb invalidation completion */
+	isb			/* Sync execution on tlb update */
 	MOV     PC, LR
```
yes

etienne-lms commented on core/arch/arm/include/kernel/tz_ssvce.h:
```diff
@@ -49,7 +49,8 @@ void arm_cl1_i_inv(void *start, void *end);
 void secure_mmu_datatlbinvall(void);
 void secure_mmu_unifiedtlbinvall(void);
-void secure_mmu_unifiedtlbinvbymva(unsigned long addr);
+void secure_mmu_unifiedtlbinvbymva_curasid(unsigned long addr);
```
I think too :)

etienne-lms commented:
Hi, since proposed to rename `secure_mmu_unifiedtlbinvXXX()` into `tlbi_XXX()` and discarding `core_tlb_maintenance()`, i will push a new proposal from scratch, at least not using incremental updates of proposed patches. Hence, this will discard previous comments on this PR.

etienne-lms commented:
Rebased and fixed from previous command.
checkpatch warm on a `asm` command.

This proposal runs ok on qemu_virt and armv7/b2260 (currently failed to test on fvp/aarch32).
But issues found on qemu_armv8: seems some TLB data are not updated. Under investigation....

jenswi-linaro commented on core/arch/arm/include/kernel/tz_ssvce.h:
```diff
@@ -48,10 +48,24 @@ void arm_cl1_i_inv_all(void);
+#ifdef ARM64
+	uint64_t mva = va;
+
+	asm volatile ("tlbi	vaae1is, %0" : : "r" (mva));
```
Please add a `write_vaae1is()` to arm64.h and use that function instead. Do the corresponding thing for ARM32 below.

jenswi-linaro commented on core/arch/arm/kernel/ssvce_a32.S:
```diff
@@ -49,80 +49,45 @@
+/* void tlbi_mva_allasid(vaddr_t mva); */
+FUNC tlbi_mva_allasid , :
 UNWIND(	.fnstart)
 
```
Empty line not needed, please fix the other functions you're touching in this file too.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -876,30 +876,48 @@ enum teecore_memtypes core_mmu_get_type_by_pa(paddr_t pa)
+		tlbi_mva_allasid_nosync(va);
+		if (size < granule)
+			break;
+		size -= granule;
```
Generally we try to avoid changing the input parameters. In this case it's not a problem, what does others think? @jforissier @jbech-linaro ?

jforissier replied:
I'm OK with this. We can make an exception for very simple functions like this one.

etienne-lms replied:
I can update the code. No issue and no drawback expected.
For my understanding, why would you rather avoid modifying content of input argument variables in a function ? Is it a matter of debugging facilities ?

jenswi-linaro replied:
No it's just that for a longer and more complex function it can become tricky to keep track of what was the input and what has been changed.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_v7.c:
```diff
@@ -642,8 +644,8 @@ void core_mmu_set_user_map(struct core_mmu_user_map *map)
 	} else {
 		write_ttbr0(read_ttbr1());
 	}
-	isb();
```
I think this `isb()` is needed, but it should perhaps be moved to inside the braces above to make it clear that it belongs to `write_contextidr()` and `write_ttbr0()` respectively.

etienne-lms replied:
Right, my fault.

etienne-lms commented:
Comments addressed... but still failing in arm64 mode (tests on arm32 mode with/without LPAE using qemu_virt are ok).

jenswi-linaro commented on core/arch/arm/include/kernel/tz_ssvce.h:
```diff
@@ -57,13 +57,11 @@ static inline void tlbi_mva_allasid_nosync(vaddr_t va)
 #ifdef ARM64
 	uint64_t mva = va;
```
This local variable assignment doesn't seem to add anything.

etienne-lms replied:
rather go for an inline explicit cast ? `tlbi_vaae1is((uintXX_t)va);`

jenswi-linaro replied:
Is the cast needed?

etienne-lms replied:
From data size, it should not, but won't strict aliasing check fail on the types ?
If current build without cast runs ok, i will update the change according to your suggestion..

jenswi-linaro commented:
It fails on FVP AArch64 pager also. Please let me know if you want help debugging.

jenswi-linaro commented:
This diff helps a bit, but obviously something more is missing since there's crashes early with xtest.

```
--- a/core/arch/arm/kernel/ssvce_a64.S
+++ b/core/arch/arm/kernel/ssvce_a64.S
@@ -39,6 +39,7 @@ END_FUNC tlbi_all
 
 /* void tlbi_mva_allasid(vaddr_t mva); */
 FUNC tlbi_mva_allasid , :
+       lsr     x0, x0, #11
        dsb     ishst           /* Sync with table update */
        tlbi    vaae1is, x0     /* Tlb by mva in inner shareable */
        dsb     ish             /* Sync with tlb invalidation completion */
```

etienne-lms commented:
Ouch! it seems you find it : bad format in argument.
The function `tlbi_asid()` is also wrong: need to left shift asid argument by 48bit.
Thanks, i'll run some more tests.

jenswi-linaro commented on core/arch/arm/include/kernel/tz_ssvce.h:
```diff
@@ -48,10 +48,18 @@ void arm_cl1_i_inv_all(void);
+static inline void tlbi_mva_allasid_nosync(vaddr_t va)
+{
+#ifdef ARM64
+	tlbi_vaae1is(va);
```
`va >> 12`

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
Rebased and R-b tag applied. Note a change versus previous proposal:
- specific commit to fix the tbli_asid() AArch64 implementation.
- include arm.h instead of arm64.h in generic tz_ssvce.h

Travis reports 2 checkpatch issues: 1 must be addressed. I will push an update.

Also: since ssvce_a32/64.S now only provides TLB maintenance, i will propose another P-R to clarify that (source file renaming ssvce_aXX.S => tlb_helpers_aXX.S).

jenswi-linaro commented:
`Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (Hikey AArch{32,64} pager)`

etienne-lms commented:
checkpatch issue fixed. T-b tag applied.
